### PR TITLE
Amesos2: Add throw for NULL numeric factorization (#5666)

### LINKED ISSUE #5666: amesos2 - add check for factorization failure.
## Enhancement
@trilinos/amesos2 

Can we add a check here 
```
       data_.numeric_ = ::KLU2::klu_factor<slu_type, local_ordinal_type>
          (sp_rowptr_with_common_type.getRawPtr(), sp_colind, sp_values,
           data_.symbolic_, &(data_.common_)) ;
      }
      else {
        data_.numeric_ = ::KLU2::klu_factor<slu_type, local_ordinal_type>
          (colptr_.getRawPtr(), rowind_.getRawPtr(), nzvals_.getRawPtr(),
           data_.symbolic_, &(data_.common_)) ;
      } //end single_process_optim_check = false

      // This is set after numeric factorization complete as pivoting can be used;
      // In this case, a discrepancy between symbolic and numeric nnz total can occur.

<<<<< HERE >>>>>
      this->setNnzLU( as<size_t>((data_.numeric_)->lnz) + as<size_t>((data_.numeric_)->unz) );

```

that is something like assert(data_.numeric_  != NULL);

I can do it, if @srajama1 or someone else can tell me the amesos2 was of throwing errors.

We have this happen to often and a segfault is less helpful

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
This would previously just seg fault on the NULL ptr.

@trilinos/<teamName>

## Motivation and Context
Provide better information if numeric factorization fails, for example if the matrix was singular.

## Checklist

- [x] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by srajama1** [comment]
Oh, can you please add your test too ?

File: packages/amesos2/src/Amesos2_KLU2_def.hpp [hunk 1]
```diff
@@ -252,6 +252,10 @@ KLU2<Matrix,Vector>::numericFactorization_impl()
            data_.symbolic_, &(data_.common_)) ;
       } //end single_process_optim_check = false
 
+      if(data_.numeric_ == NULL) {
+          throw std::runtime_error("KLU2 numeric factorization failed.");
```
Comment (mhoemmen) [inline packages/amesos2/src/Amesos2_KLU2_def.hpp]: `TEUCHOS_TEST_FOR_EXCEPTION` can give better information, like caller info, a hook for setting a debugger break point, and an optional stack trace.
```c++
TEUCHOS_TEST_FOR_EXCEPTION
  (data_.numeric_ == nullptr, std::runtime_error,
   "KLU2 numeric factorization failed");
```

Comment (MicheldeMessieres) [reply packages/amesos2/src/Amesos2_KLU2_def.hpp]: Updated. Thanks.

File: packages/amesos2/test/solvers/KLU2_UnitTests.cpp [hunk 2]
```diff
@@ -271,6 +271,48 @@ namespace {
     }
   }
 
+  TEUCHOS_UNIT_TEST_TEMPLATE_3_DECL( KLU2, NumericFactorizationNullThrows, SCALAR, LO, GO )
+  {
+    typedef ScalarTraits<SCALAR> ST;
+    typedef CrsMatrix<SCALAR,LO,GO,Node> MAT;
+    typedef MultiVector<SCALAR,LO,GO,Node> MV;
+    //typedef KLU2<MAT,MV> SOLVER;
```
Comment (mhoemmen) [inline packages/amesos2/test/solvers/KLU2_UnitTests.cpp]: It's OK to delete lines of code :-)

Comment (MicheldeMessieres) [reply packages/amesos2/test/solvers/KLU2_UnitTests.cpp]: Removed. Thanks.

**Comment by srajama1** [comment]
@MicheldeMessieres : Thanks !

**Pull request merged**